### PR TITLE
[REVIEW] UCX/UCX-Py Debugging/Performance checks

### PR DESCRIPTION
**Pull request by quasiben**
Very early work on writing down all the diagnostics we have utilized in uncovering performance issues and/or configurations issues

### CODE REVIEW

File: docs/source/ucx-debug.rst [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+
+ucx_perftest
+------------
+
+Perf test for IB should be in the 10+ GB/s range
+::
+
+    CUDA_VISIBLE_DEVICES=0 UCX_NET_DEVICES=mlx5_0:1 UCX_TLS=rc,cuda_copy ucx_perftest -t tag_bw -m cuda -s 10000000 -n 10 -p 9999 & \
+    CUDA_VISIBLE_DEVICES= 1UCX_NET_DEVICES=mlx5_1:1 UCX_TLS=rc,cuda_copy ucx_perftest `hostname` -t tag_bw -m cuda -s 100000000 -n 10 -p 9999
```
Comment (pentschev) [inline docs/source/ucx-debug.rst:26]:
```suggestion
    CUDA_VISIBLE_DEVICES=1 UCX_NET_DEVICES=mlx5_1:1 UCX_TLS=rc,cuda_copy ucx_perftest `hostname` -t tag_bw -m cuda -s 100000000 -n 10 -p 9999
```

File: docs/source/ucx-debug.rst [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    | Message size: 100000000                                                                  |
+    +------------------------------------------------------------------------------------------+
+                10     0.000  9104.800  9104.800   10474.41   10474.41         110         110
+
+
+Perf test for NVLink should be in the 20+ GB/s range
+::
+
+    CUDA_VISIBLE_DEVICES=0 UCX_TLS=cuda_ipc,cuda_copy,tcp,sockcm UCX_SOCKADDR_TLS_PRIORITY=sockcm  ucx_perftest -t tag_bw -m cuda -s 10000000 -n 10 -p 9999 -c 0 & \
+    CUDA_VISIBLE_DEVICES=1 UCX_TLS=cuda_ipc,cuda_copy,tcp,sockcm UCX_SOCKADDR_TLS_PRIORITY=sockcm ucx_perftest `hostname` -t tag_bw -m cuda -s 100000000 -n 10 -p 9999 -c 1
```
Comment (pentschev) [inline docs/source/ucx-debug.rst:48]: `-c` is NUMA dependent, so you should be explicit in specifying that the user must choose a CPU that matches the CPU Affinity for that GPU, as reported by `nvidia-smi topo -m`.

**Review by pentschev — CHANGES_REQUESTED** [review]

**Comment by quasiben** [comment]
I think this is ready for another pass if someone has time.  

What do folks think about adding additional `ucx_perftest` examples ?

File: docs/source/ucx-debug.rst [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+      NV#  = Connection traversing a bonded set of # NVLinks
+
+NVLink
+------
+
+System Configuration
+~~~~~~~~~~~~~~~~~~~~
+
+Again, using the ``nvidia-smi topo -m`` command which show the hardware topology of the GPUs and display NVLink information.  For example, using the output above, we can see that GPUs 0-4 are conneted to another with NVLink,
+and GPUs 5-8 are also connected with NVLink.  However, these sets of GPUs are only connected to one another with InfiniBand.
```
Comment (pentschev) [inline docs/source/ucx-debug.rst:85]: The description you wrote for NVLink is incorrect, please refer to https://images.nvidia.com/content/pdf/dgx1-v100-system-architecture-whitepaper.pdf (PDF page 11/document page 9). We can't say some GPUs are connected to "one" NVLink and other GPUs are connected to "another" NVLink, and connectivity is not homogenous. What happens is we have GPUs connected that are connected by a single NVLink connection (NV1), others with two NVLink connections (NV2), and some not connected at all via NVLink (SYS).

Can you please rewrite that? My suggestion would be something like:

"Note that NVLink connectivity on the system above (DGX-1) is not homogenous, some GPUs are connected by a single NVLink connection (NV1, e.g., GPUs 0 and 1), others with two NVLink connections (NV2, e.g., GPUs 1 and 2), and some not connected at all via NVLink (SYS, e.g., GPUs 3 and 4)."

File: docs/source/ucx-debug.rst (hunk 2 above)
Comment (pentschev) [reply docs/source/ucx-debug.rst:95]: This is resolved in https://github.com/rapidsai/ucx-py/pull/624/commits/8eb2d479ed94627ce7cb13f36a1df55871ff8a7c .

File: docs/source/ucx-debug.rst [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    | Data layout:  (automatic)                                                                |
+    | Send memory:  cuda                                                                       |
+    | Recv memory:  cuda                                                                       |
+    | Message size: 100000000                                                                  |
+    +------------------------------------------------------------------------------------------+
+                10     0.000  4163.694  4163.694   22904.52   22904.52         240         240
+
+
+Experiental Debugging
+---------------------
```
Comment (pentschev) [inline docs/source/ucx-debug.rst:113]:
```suggestion
Experimental Debugging
----------------------
```

File: docs/source/ucx-debug.rst [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    | Test:         tag match bandwidth                                                        |
+    | Data layout:  (automatic)                                                                |
+    | Send memory:  cuda                                                                       |
+    | Recv memory:  cuda                                                                       |
+    | Message size: 100000000                                                                  |
+    +------------------------------------------------------------------------------------------+
+                10     0.000  9104.800  9104.800   10474.41   10474.41         110         110
+
+
+``-c`` option is NUMA dependent and sets the CPU Affinity of process for a partiular GPU.  CPU Affinity information can be found in ``nvidia-smi topo -m``
```
Comment (pentschev) [inline docs/source/ucx-debug.rst:50]:
```suggestion
``-c`` option is NUMA dependent and sets the CPU Affinity of process for a particular GPU.  CPU Affinity information can be found in ``nvidia-smi topo -m``
```

File: docs/source/ucx-debug.rst [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+::
+
+    user@mlnx:~$ ibdev2netdev
+    mlx5_0 port 1 ==> ib0 (Up)
+    mlx5_1 port 1 ==> ib1 (Up)
+    mlx5_2 port 1 ==> ib2 (Up)
+    mlx5_3 port 1 ==> ib3 (Up)
+
+``ucx_info -d`` and ``ucx_info -p -u t`` are helpful commands to display what UCX understand about the underlying hardware
```
Comment (pentschev) [inline docs/source/ucx-debug.rst:21]:
```suggestion
``ucx_info -d`` and ``ucx_info -p -u t`` are helpful commands to display what UCX understands about the underlying hardware
```

**Review by pentschev — CHANGES_REQUESTED** [review]
@quasiben I added a few typo fixes and requested some changes regarding your NVLink description, please take a look when you have a chance.

File: docs/source/ucx-debug.rst (hunk 3 above)
Comment (quasiben) [reply docs/source/ucx-debug.rst:85]: Thanks Peter!

**Comment by quasiben** [comment]
Thanks again @pentschev -- used all of your suggestions in the last commit

File: docs/source/ucx-debug.rst [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    | Data layout:  (automatic)                                                                |
+    | Send memory:  cuda                                                                       |
+    | Recv memory:  cuda                                                                       |
+    | Message size: 100000000                                                                  |
+    +------------------------------------------------------------------------------------------+
+                10     0.000  4163.694  4163.694   22904.52   22904.52         240         240
+
+
+ExperimentalDebugging
+---------------------
```
Comment (pentschev) [inline docs/source/ucx-debug.rst:116]:
```suggestion
Experimental Debugging
----------------------
```

Comment (quasiben) [reply docs/source/ucx-debug.rst]: Oops

**Comment by jakirkham** [comment]
Thanks Ben! Also thanks for reviewing Peter! 😄

**Pull request merged**